platform/pwl.c

```diff
@@ -39,3 +61,775 @@ pwl_display_destroy (PwlDisplay *self)
+
+    g_debug ("New XDG toplevel configuration: (%" PRIu32 ", %" PRIu32 ")", width, height);
+
+    wl_data.resize_window(data);
+}
+
+static void
+xdg_toplevel_on_close (void *data, struct xdg_toplevel *xdg_toplevel)
+{
+    // TODO g_application_quit (g_application_get_default ());
+}
+
+static const struct xdg_toplevel_listener xdg_toplevel_listener = {
+    .configure = xdg_toplevel_on_configure,
+    .close = xdg_toplevel_on_close,
+};
+
+
+gboolean create_window (void *data, GError **error)
+{
+    g_debug ("Creating Wayland surface...");
+
+    win_data.wl_surface = wl_compositor_create_surface (wl_data.compositor);
+    g_assert (win_data.wl_surface);
+
+#if HAVE_DEVICE_SCALING
+    wl_surface_add_listener (win_data.wl_surface, &win_data.surface_listener, data);
+#endif /* HAVE_DEVICE_SCALING */
+
+    if (wl_data.xdg_shell != NULL) {
+        win_data.xdg_surface =
+            xdg_wm_base_get_xdg_surface (wl_data.xdg_shell,
+                                         win_data.wl_surface);
+        g_assert (win_data.xdg_surface);
+
+        xdg_surface_add_listener (win_data.xdg_surface, &xdg_surface_listener,
+                                  NULL);
+        win_data.xdg_toplevel =
+            xdg_surface_get_toplevel (win_data.xdg_surface);
+        g_assert (win_data.xdg_toplevel);
+
+        xdg_toplevel_add_listener (win_data.xdg_toplevel,
+                                   &xdg_toplevel_listener, data);
+        // TODO: xdg_toplevel_set_title (win_data.xdg_toplevel, COG_DEFAULT_APPNAME);
+
+        const char *app_id = NULL;
+        GApplication *app = g_application_get_default ();
+        if (app) {
+            app_id = g_application_get_application_id (app);
+        }
+        if (!app_id) {
+            // TODO: app_id = COG_DEFAULT_APPID;
+        }
+        xdg_toplevel_set_app_id (win_data.xdg_toplevel, app_id);
+        wl_surface_commit(win_data.wl_surface);
+    } else if (wl_data.fshell != NULL) {
+        zwp_fullscreen_shell_v1_present_surface (wl_data.fshell,
+                                                 win_data.wl_surface,
+                                                 ZWP_FULLSCREEN_SHELL_V1_PRESENT_METHOD_DEFAULT,
+                                                 NULL);
+    } else if (wl_data.shell != NULL) {
+        win_data.shell_surface = wl_shell_get_shell_surface (wl_data.shell,
+                                                             win_data.wl_surface);
+        g_assert (win_data.shell_surface);
+
+        wl_shell_surface_add_listener (win_data.shell_surface,
+                                       &shell_surface_listener,
+                                       data);
+        wl_shell_surface_set_toplevel (win_data.shell_surface);
+
+        /* wl_shell needs an initial surface configuration. */
+        configure_surface_geometry (0, 0);
+    }
+
+    const char* env_var;
+    if ((env_var = g_getenv ("COG_PLATFORM_FDO_VIEW_FULLSCREEN")) &&
```

Comment:
We *need* to stop using environment variables here. The idea is to have a `pwl_window_set_fullscreen(PwlWindow*, bool)` function that can be called, and a virtual method `cog_view_set_fullscreen(CogView*, gboolean)` + `CogView.is-fullscreen` property.